### PR TITLE
poetry: support for poetry metadata format 2.0

### PR DESCRIPTION
**Pull request by GabMus**

### CODE REVIEW

**Review by ENDrain — APPROVED** [review]
LGTM. Maybe could also add a line in README.md listing metadata formats supported, but not sure if that's necessary. @mig5, since this worked for you could you also please add an approving review?

**Comment by GabMus** [comment]
I really think we should merge this

**Comment by d4nj1** [comment]
Would also be nice if the `--yaml` output format could be added similar to

https://github.com/flatpak/flatpak-builder-tools/blob/master/pip/flatpak-pip-generator#L55-L67
with
https://github.com/flatpak/flatpak-builder-tools/blob/master/pip/flatpak-pip-generator#L480-L491

**Comment by hfiguiere** [comment]
Can you rebase this? We could get it merged eventually.

**Comment by uninstall-your-browser** [comment]
`--production` does not seem to work, it still tries (and fails) to retrieve dev dependencies (unless I'm somehow generating my lock file wrong?). It should probably be changed to `--without <group-name>`, since poetry has deprecated `--no-dev`.

**Comment by hfiguiere** [comment]
@GabMus unless this gets rebased, this can't be merged.

**Comment by GabMus** [comment]
rebased, note that this PR is ancient and I don't even use flatpak-builder-tools anymore, so make sure this works as expected before merging

File: poetry/flatpak-poetry-generator.py [hunk 1]
```diff
@@ -63,19 +63,23 @@ def get_module_sources(parsed_lockfile: dict, include_devel: bool = True) -> lis
...
+                    (
+                        package.get("category") == "dev"
+                        and include_devel
+                        and not package.get("optional")
+                    )
+                    or (
+                        package.get("category") == "main"
+                        and not package.get("optional")
+                    )
                 ):
```
Comment (bbhtt) [inline poetry/flatpak-poetry-generator.py:76]: the original logic and the rewritten one can be simplified a bit I think. It's a bit hard to follow this many ands/ors. Something like

```
category = package.get("category")
optional = package.get("optional", False)

if (
    not category
    or (category == "dev" and include_devel and not optional)
    or (category == "main" and not optional)
):
```

**Comment by bbhtt** [comment]
> Would also be nice if the --yaml output format could be added similar to

This should be a separate issue. You just include the generated manifest. Having multiple formats is not a top priority.

> `--production` does not seem to work, it still tries (and fails) to retrieve dev dependencies

It should work now for v2 metadata.

>  Though it should probably be changed to `--without <group-name>`, since poetry has deprecated `--no-dev`.

Indeed, generic group support can be done later. This can be a separate issue. For now `--production` excludes things that have only `dev` in their groups.

File: poetry/flatpak-poetry-generator.py (hunk 1 above)
Comment (bbhtt) [reply poetry/flatpak-poetry-generator.py]: Applied

**Comment by bbhtt** [comment]
Tested this on a v1 and v2 lock files with the `--production` combination. Seems to work.

[poetryv2.txt](https://github.com/user-attachments/files/20025837/poetryv2.txt)
[poetryv1.txt](https://github.com/user-attachments/files/20025838/poetryv1.txt)
